GitUI/CommandsDialogs/FormCommit.cs

```diff
@@ -1902,16 +1929,41 @@ private void Stage(IReadOnlyList<GitItemStatus> items)
                         }
 
                         unstagedFiles.RemoveAll(item => !item.IsSubmodule && unstagedItems.Contains(item));
+
                         unstagedFiles.RemoveAll(
-                            item => item.IsSubmodule && item.GetSubmoduleStatusAsync().IsCompleted &&
-                                    (item.GetSubmoduleStatusAsync().CompletedResult() is null ||
-                                     (!item.GetSubmoduleStatusAsync().CompletedResult().IsDirty && unstagedItems.Contains(item))));
-                        foreach (var item in unstagedItems.Where(
-                            item => item.IsSubmodule &&
-                                    (ThreadHelper.JoinableTaskFactory.Run(() => item.GetSubmoduleStatusAsync()) is null ||
-                                     (item.GetSubmoduleStatusAsync().IsCompleted && item.GetSubmoduleStatusAsync().CompletedResult().IsDirty))))
+                            item =>
+                            {
+                                if (!item.IsSubmodule)
+                                {
+                                    return false;
+                                }
+
+                                Task<GitSubmoduleStatus?> statusTask = item.GetSubmoduleStatusAsync();
+
+                                if (!statusTask.IsCompleted)
+                                {
+                                    return false;
+                                }
+
+                                GitSubmoduleStatus? status = statusTask.CompletedResult();
+                                return status is null || (!status.IsDirty && unstagedItems.Contains(item));
+                            });
+
+                        foreach (var item in unstagedItems)
                         {
-                            item.GetSubmoduleStatusAsync().CompletedResult().Status = SubmoduleStatus.Unknown;
+                            if (!item.IsSubmodule)
+                            {
+                                continue;
+                            }
+
+                            GitSubmoduleStatus? gitSubmoduleStatus = ThreadHelper.JoinableTaskFactory.Run(() => item.GetSubmoduleStatusAsync());
+
+                            if (gitSubmoduleStatus is null || !gitSubmoduleStatus.IsDirty)
+                            {
+                                continue;
+                            }
+
+                            gitSubmoduleStatus.Status = SubmoduleStatus.Unknown;
```

Comment:
There was a NRE bug in here, but I can't see it now. The previous code was pretty hard to unpack. Changed to be more verbose, but clearer. This form also allows the compiler to validate null value flow.